docs/knowledgebase/runtime/macros.md

```diff
@@ -2,58 +2,268 @@
+
+**Doc. and Example**
+
+  - [API Documentation](https://substrate.dev/rustdocs/v2.0.0-rc3/frame_support/macro.decl_error.html)
+  - [Macro Definition](https://github.com/paritytech/substrate/blob/v2.0.0-rc3/frame/support/src/error.rs#L71-L203)
+  - Macro Expansion Example: [original](https://gist.github.com/jimmychu0807/c4a88ec8e0342ee9f4e14bd26287324e#file-pallet-template-rs-L53-L60), [expanded](https://gist.github.com/jimmychu0807/c4a88ec8e0342ee9f4e14bd26287324e#file-pallet-template-expanded-rs-L331-L417)
+
+**Macro Implementation Notes**
+
+  - `Error<T: Trait>` enum type is filled with enum variants specified in the macro.
+  - `Error` implements various helper functions, and maps each error to a sequential error code and an error string.
+  - `Error` type implementation contains a `fn metadata()` function.
+  - `Error` implements `From<Error<T>>` for `DispatchError`, so the error type can be returned in dispatchable functions.
+
+### decl_module!
+
+**When to Use**
+
+To define dispatchable functions in a pallet
+
+**What It Does**
+
+The macro declares a `Module` struct and `Call` enum type for the containing pallet. It combines the necessary logics using user-defined dispatchable calls into the two types. In addition to various helper traits implemented for `Module` and `Call`, e.g. `Copy`, `StructuralEq`, `Debug`, the macro also inserts lifecycle trait implementations for `Module`, e.g. `frame_support::traits::OnInitialize`, `frame_support::traits::OnFinalize`, `frame_support::traits::OnRuntimeUpgrade`, and `frame_support::traits::OffchainWorker`.
+
+**Doc. and Example**
+
+  - [API Documentation](https://substrate.dev/rustdocs/v2.0.0-rc3/frame_support/macro.decl_module.html)
+  - [Macro Definition](https://github.com/paritytech/substrate/blob/v2.0.0-rc3/frame/support/src/dispatch.rs#L275-L1613)
+  - Macro Expansion Example: [original](https://gist.github.com/jimmychu0807/c4a88ec8e0342ee9f4e14bd26287324e#file-pallet-template-rs-L63-L109), [expanded](https://gist.github.com/jimmychu0807/c4a88ec8e0342ee9f4e14bd26287324e#file-pallet-template-expanded-rs-L419-L971)
+
+**Macro Implementation Notes**
+
+  - This macro declares a `Module<T>` struct and a `Call<T>` enum which implements the dispatch logic.
+  - Helper traits are automatically imeplemented for `Module` struct, such as `core::cmp::Eq`, `core::clone::Clone`, etc.
+  - `Module` implements various lifecycle traits, such as `frame_support::traits::OnInitialize` (callback when a block is initialized), `frame_support::traits::OnRuntimeUpgrade` (callback when a chain is undergoing a runtime upgrade), `frame_support::traits::OnFinalize` (callback when a block is finalized), and `frame_support::traits::OffchainWorker` (entrance of an offchain worker upon a block is finalized). 
+  - `Call<T: Trait>` implements `frame_support::dispatch::GetDispatchInfo` and `frame_support::dispatch::Dispatchable` traits that contain the core logic for dispatching calls.
```

Comment:
actually since in-origin filtering the call only implements https://substrate.dev/rustdocs/v2.0.0-rc4/frame_support/traits/trait.UnfilteredDispatchable.html

(because module call can't convert itself to outer call easily, and the filter parameter is the outer call)